### PR TITLE
event and tracker implementation

### PR DESCRIPTION
**Pull request by xavluiz**
tracker and event implementation logic

### CODE REVIEW

File: SoftwareCo/SoftwareCo/Managers/DocEventManager.cs [hunk 1]
```diff
@@ -180,7 +170,8 @@ public async void DocEventsOnDocumentOpenedAsync(Document document)
             InitPluginDataIfNotExists();
             _pluginData.InitFileInfoIfNotExists(fileName);
 
-            /**
+            TrackerUtilManager.TrackEditorFileActionEvent("file", "open", fileName);
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Managers/DocEventManager.cs:173]: you have a handle to the Document here, why not pass that to the Tracker itself?

Document has these properties:
https://docs.microsoft.com/en-us/dotnet/api/envdte.document?view=visualstudiosdk-2017#properties

File: SoftwareCo/SoftwareCo/Managers/ProjectManager.cs [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+                        fd.syntax = projItem.Document.Language;
+                    }
+                }
+                catch (Exception e)
+                {
+                    Logger.Info($"Unable to obtain file language: {e.Message}");
+                }
+            } else
+            {
+                fd.project_name = "Unnamed";
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Managers/ProjectManager.cs:62]: @bjacobson26 does this line up with the unsaved new files from sublimeText?

File: SoftwareCo/SoftwareCo/Managers/TrackerUtilManager.cs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+                codetimeEvent.projectEntity = await GetProjectEntity(fileInfo.file);
+
+                if (repoEntity == null)
+                {
+                    repoEntity = GetRepoEntity(codetimeEvent.projectEntity.project_directory);
+                }
+
+                codetimeEvent.repoEntity = repoEntity;
+
+                Console.WriteLine($"CODE TIME EVENT: {codetimeEvent.ToString()}");
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Managers/TrackerUtilManager.cs:54]: we can remove this after testing

File: SoftwareCo/SoftwareCo/Managers/ProjectManager.cs [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+    class ProjectManager
+    {
+        public static DTE2 ObjDte { set; get; }
+        public static string solutionDirectory { set; get; }
+
+        /**
+         * This will return the project and file detail info.
+         * A fileName does not have to be passed in to fetch the current project.
+         **/
+        public static async Task<FileDetails> GetFileDatails(string fileName)
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Managers/ProjectManager.cs:19]: typo in name.   should be `GetFileDetails`.

File: SoftwareCo/SoftwareCo/Managers/TrackerUtilManager.cs [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+            PluginEntity pluginEntity = new PluginEntity();
+            pluginEntity.plugin_version = SoftwareCoPackage.GetVersion();
+            pluginEntity.plugin_id = Constants.PluginId;
+            pluginEntity.plugin_name = Constants.PluginName;
+            return pluginEntity;
+        }
+
+        public static async Task<ProjectEntity> GetProjectEntity(string fileName)
+        {
+            FileDetails fd = await ProjectManager.GetFileDatails(fileName);
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Managers/TrackerUtilManager.cs:125]: typo in name. should be GetFileDetails.

File: SoftwareCo/SoftwareCo/Managers/TrackerUtilManager.cs [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+            repoEntity.git_tag = info.tag;
+            repoEntity.owner_id = info.ownerId;
+            repoEntity.repo_identifier = info.identifier;
+            repoEntity.repo_name = info.repoName;
+            return repoEntity;
+        }
+
+        public static async Task<FileEntity> GetFileEntity(string fileName)
+        {
+            FileDetails fd = await ProjectManager.GetFileDatails(fileName);
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Managers/TrackerUtilManager.cs:146]: typo in name. should be GetFileDetails.

File: SoftwareCo/SoftwareCo/SoftwareCoPackage.cs [hunk 7]
```diff
@@ -271,11 +262,6 @@ public async void SolutionEventOpenedAsync()
                     CodeMetricsTreeManager.Instance.OpenCodeMetricsPaneAsync();
                 }
 
-                Task.Delay(3000).ContinueWith((task) =>
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/SoftwareCoPackage.cs:274]: should the new  editor activate event get called here still?

File: SoftwareCo/SoftwareCo/SoftwareOpenCodeMetricsTreeCommand.cs [hunk 8]
```diff
@@ -87,7 +87,6 @@ private static void Execute(object sender, EventArgs e)
             try
             {
                 CodeMetricsTreeManager.Instance.OpenCodeMetricsPaneAsync();
-                EventManager.Instance.CreateCodeTimeEvent("mouse", "click", "ShowTreeView");
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/SoftwareOpenCodeMetricsTreeCommand.cs:90]: we should still track this event.   but with the correct event names and locations.     @bjacobson26 added one for this in Sublime

File: SoftwareCo/SoftwareCo/Tracker/client/Http.cs [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+        public static async Task<Response> GetAsync(string api)
+        {
+            try
+            {
+                using (HttpClient client = new HttpClient())
+                {
+                    client.Timeout = TimeSpan.FromSeconds(15);
+                    if (CacheManager.HasJwt())
+                    {
+                        client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Authorization", CacheManager.jwt);
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/client/Http.cs:29]: lets add the version headers in here

File: SoftwareCo/SoftwareCo/Tracker/client/Http.cs [hunk 10]
```diff
@@ -0,0 +1,91 @@
...
+        public static async Task<Response> PostAsync(string api, object payload)
+        {
+            try
+            {
+                using (HttpClient client = new HttpClient())
+                {
+                    client.Timeout = TimeSpan.FromSeconds(15);
+                    if (CacheManager.HasJwt())
+                    {
+                        client.DefaultRequestHeaders.Authorization = new AuthenticationHeaderValue("Authorization", CacheManager.jwt);
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/client/Http.cs:58]: version headers here too please

File: SoftwareCo/SoftwareCo/Tracker/client/Http.cs [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+                    // create the url
+                    string url = $"{Constants.api_endpoint}{api}";
+
+                    // make the POST call
+                    string payloadStr = JsonConvert.SerializeObject(payload);
+                    HttpContent content = new StringContent(payloadStr, Encoding.UTF8, "application/json");
+                    HttpResponseMessage resp = await client.PostAsync(url, content);
+                    resp.EnsureSuccessStatusCode();
+
+                    return await BuildResponse(resp);
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/client/Http.cs:70]: do we need to build this into a custom response?   wouldn't the HTTPResponseMessage object be enough?

File: SoftwareCo/SoftwareCo/Tracker/events/CodetimeEvent.cs [hunk 12]
```diff
@@ -0,0 +1,58 @@
...
+        public PluginEntity pluginEntity { get; set; }
+        public ProjectEntity projectEntity { get; set; }
+        public RepoEntity repoEntity { get; set; }
+
+        public GenericContext buildContext()
+        {
+            GenericContext context = new GenericContext()
+                .SetSchema("iglu:com.software/codetime/jsonschema/1-0-1")
+                .Add("keystrokes", keystrokes)
+                .Add("keystrokes", keystrokes)
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/events/CodetimeEvent.cs:30]: keystrokes is added twice

File: SoftwareCo/SoftwareCo/Tracker/managers/CacheManager.cs [hunk 13]
```diff
@@ -0,0 +1,32 @@
+﻿using System;
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/managers/CacheManager.cs:1]: are we using this cacheManager or the one at root?

File: SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs [hunk 14]
```diff
@@ -0,0 +1,76 @@
+﻿using System;
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs:1]: which HashManager are we using?

File: SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+        {
+            if (!CacheManager.HasJwt() || string.IsNullOrEmpty(value))
+            {
+                return "";
+            }
+
+            try
+            {
+                byte[] key = GenericHash.GenerateKey(); // 64 byte key
+                string hashedValue = Utilities.BinaryToHex(GenericHash.Hash(value, key, 128));
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs:20]: I'm pretty sure that if we use keys, our hashes for the same value become random.    Did you check that running this twice on the same original value will produce the same hashed value?

File: SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs [hunk 16]
```diff
@@ -0,0 +1,76 @@
...
+                if (CacheManager.HasCachedValue(dataType, hashedValue))
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs:21]: I think the logic is backward here.   Should be `if (CacheManager.DoesNotHaveCachedValue(dataType, hashedValue))`

where `CacheManager.DoesNotHaveCachedValue` == `!CacheManager.HasCachedValue`

File: SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs [hunk 17]
```diff
@@ -0,0 +1,76 @@
...
+            catch (Exception e)
+            {
+                Console.WriteLine("SwdcVsTracker - Hash value error: {0}", e.Message);
+            }
+            return "";
+        }
+
+        private static async void EncryptValue(string value, string hashedValue, string dataType)
+        {
+            UserEncryptedData data = new UserEncryptedData();
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs:40]: do we really need the UserEncryptedData object?   Can we just put the data in the call to PostAsync directly?

File: SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs [hunk 18]
```diff
@@ -0,0 +1,76 @@
...
+            data.data_type = dataType;
+            data.hashed_value = hashedValue;
+            data.value = value;
+
+            await Http.PostAsync("/user_encrypted_data", data);
+        }
+
+        private static async void PopulateHashValues()
+        {
+            Response resp = await Http.GetAsync("/hashed_values");
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs:50]: why use the GetAsync vs a blocking Get if we are going to `await` on it anyway?   Is there only the Async one?

File: SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs [hunk 19]
```diff
@@ -0,0 +1,76 @@
...
+            if (resp.ok)
+            {
+                // serialize the object to the UserHashedValues object
+                string json = JsonConvert.SerializeObject(resp.responseData);
+                try
+                {
+                    Dictionary<string, List<string>> dictionary = JsonConvert.DeserializeObject<Dictionary<string, List<string>>>(json);
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs:57]: seems like there could be a simpler way than to serialize and then deserialize this object.  But I don't know C sharp

File: SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs [hunk 20]
```diff
@@ -0,0 +1,76 @@
...
+
+                    // go through each key and set the cache
+                    foreach (string key in dictionary.Keys)
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs:60]: Actually.   you can just push the response from hashed_values right into the cache.   It is the cache.   This is really a reset of the cache, vs a new entry.

File: SoftwareCo/SoftwareCo/Tracker/managers/TrackerManager.cs [hunk 21]
```diff
@@ -0,0 +1,94 @@
...
+        {
+            // initialie our http client with the endpoint that fetches the snowplow collector endpoint
+            Http.Initialize(swdcApiHost);
+
+            // fetch the tracker_api from the plugin config
+            Response resp = await Http.GetAsync("/plugins/config");
+
+            if (resp.ok && resp.responseData != null)
+            {
+                string json = JsonConvert.SerializeObject(resp.responseData);
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/managers/TrackerManager.cs:43]: is this two step process standard for C sharp?

File: SoftwareCo/SoftwareCo/Tracker/managers/TrackerManager.cs [hunk 22]
```diff
@@ -0,0 +1,94 @@
...
+                // Controls the storage of events
+                // NOTE: You must dispose of storage yourself when closing your application!
+                LiteDBStorage storage = new LiteDBStorage("events.db");
+
+                // Controls queueing events
+                PersistentBlockingQueue queue = new PersistentBlockingQueue(storage, new PayloadToJsonString());
+
+                // Controls pulling events of the queue and pushing them to the sender
+                AsyncEmitter emitter = new AsyncEmitter(endpoint, queue);
+                Subject subject = new Subject().SetPlatform(Platform.Iot).SetLang("EN");
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/managers/TrackerManager.cs:59]: "EN" isn't necessarily correct here.    we should pick that up from the OS.   But let's leave that off for now

File: SoftwareCo/SoftwareCo/Tracker/managers/TrackerManager.cs [hunk 23]
```diff
@@ -0,0 +1,94 @@
...
+                t = Tracker.Instance;
+
+                t.Start(emitter, subject, null, trackerNamespace, appId, false /*encodeBase64*/, false /*synchronous*/);
+            }
+        }
+
+        public void TrackCodetimeEvent(CodetimeEvent codetimeEvent)
+        {
+            CacheManager.jwt = codetimeEvent.authEntity != null ? codetimeEvent.authEntity.jwt : "";
+            SelfDescribing selfDescribing = codetimeEvent.buildContexts();
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/Tracker/managers/TrackerManager.cs:70]: do these methods need to call `sendEvent`?

File: SoftwareCo/SoftwareCo/TrackerManager.cs [hunk 24]
```diff
@@ -0,0 +1,94 @@
+﻿using System.Collections.Generic;
```
Comment (mmclead) [inline SoftwareCo/SoftwareCo/TrackerManager.cs:1]: we have 2 TrackerManagers.    Which one is being used???

File: SoftwareCo/SoftwareCo/CodeMetricsTree.xaml.cs [hunk 25]
```diff
@@ -485,19 +485,30 @@ private void EmailConnectClickHandler(object sender, MouseButtonEventArgs args)
...
         {
             SoftwareCoUtil.launchLogin(loginType);
-            EventManager.Instance.CreateCodeTimeEvent("mouse", "click", "LaunchLoginOnboard");
         }
 
         private void LaunchWebDashboard(object sender, MouseButtonEventArgs args)
         {
             SoftwareCoUtil.launchWebDashboard();
-            EventManager.Instance.CreateCodeTimeEvent("mouse", "click", "LaunchWebDashboard");
+            UIElementEntity entity = new UIElementEntity();
```
[the remark this answers is not in the document]
Comment (mmclead) [reply SoftwareCo/SoftwareCo/CodeMetricsTree.xaml.cs:493]: never mind.  found those.

File: SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs (hunk 15 above)
Comment (pbstriker38) [reply SoftwareCo/SoftwareCo/Tracker/managers/HashManager.cs:20]: Yeah. Key would make it different on every run and different for users on the same team.

**Pull request closed without merging**